### PR TITLE
Re-configuring kube-proxy to not be a privileged container

### PR DESCRIPTION
**Pull request by kevensen**
I recommend not running the kube-proxy as a privileged container.  

**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Currently the kube-proxy is run as a privileged container.  A privileged container can be a vector to obtaining host (node) access.  By removing the `privileged` clause in the security context and adding the `NET_ADMIN` Linux capability, the kube-proxy can preserve its function and prevent the mounting of cgroups in the container.   

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: cluster/addons/kube-proxy/kube-proxy-ds.yaml [hunk 1]
```diff
@@ -48,7 +48,9 @@ spec:
         {{kube_cache_mutation_detector_env_name}}
           {{kube_cache_mutation_detector_env_value}}
         securityContext:
-          privileged: true
+          capabiltiies:
+            add:
+            - NET_ADMIN
```
Comment (neolit123) [inline cluster/addons/kube-proxy/kube-proxy-ds.yaml:53]: something tells me that this isn't sufficient, have you tested the change in all kube-proxy modes?
AFAIK, `NET_RAW` should be required for "iptables" too.

@kubernetes/sig-network-pr-reviews

Comment (neolit123) [reply cluster/addons/kube-proxy/kube-proxy-ds.yaml:53]: the `iptables -w` option requires mounting `/run/xtables.lock` which needs root.
https://github.com/kubernetes/kubernetes/issues/46103
https://github.com/kubernetes/kubernetes/pull/46372

Comment (kevensen) [reply cluster/addons/kube-proxy/kube-proxy-ds.yaml:53]: IIRC, the UID in the container is 0.  Please remind me, mounting a host path shouldn't need CAP_SYS_ADMIN.

To answer your earlier question, no, I have not tested this in all kube-proxy modes.  Was having a little trouble getting me dev env setup.  I'll try harder and test more locally.

**Comment by kevensen** [comment]
I signed it

File: cluster/addons/kube-proxy/kube-proxy-ds.yaml (hunk 1 above)
Comment (neolit123) [reply cluster/addons/kube-proxy/kube-proxy-ds.yaml:53]: > IIRC, the UID in the container is 0. Please remind me, mounting a host path shouldn't need CAP_SYS_ADMIN.

it shouldn't. ok, i guess the mounting is not a problem here.

> To answer your earlier question, no, I have not tested this in all kube-proxy modes. Was having a little trouble getting me dev env setup. I'll try harder and test more locally.

if this PR gets stuck without review, you could ask for a discussion in #sig-network on k8s slack.

Comment (kevensen) [reply cluster/addons/kube-proxy/kube-proxy-ds.yaml:53]: cool, thanks!

**Comment by neolit123** [comment]
since i'm curious about the change as is, let's run it in CI to see what happens.
i believe iptables is used for both deployers that accept the changed specs (kubeadm / kube-up).

File: cmd/kubeadm/app/phases/addons/proxy/manifests.go [hunk 2]
```diff
@@ -80,7 +80,9 @@ spec:
         - --config=/var/lib/kube-proxy/{{ .ProxyConfigMapKey }}
         - --hostname-override=$(NODE_NAME)
         securityContext:
-          privileged: true
+	  capabilities:
+	    add:
+	    - NET_ADMIN
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/addons/proxy/manifests.go:85]: this seems to be failing with:

```
I1106 00:14:37.172875     175 round_trippers.go:444] POST https://kind-control-
yaml: line 30: found character that cannot start any token
unable to decode kube-proxy daemonset
k8s.io/kubernetes/cmd/kubeadm/app/phases/addons/proxy.createKubeProxyAddon
	cmd/kubeadm/app/phases/addons/proxy/proxy.go:145
k8s.io/kubernetes/cmd/kubeadm/app/phases/addons/proxy.EnsureProxyAddon
	cmd/kubeadm/app/phases/addons/proxy/proxy.go:58
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/init.runKubeProxyAddon
	cmd/kubeadm/app/cmd/phases/init/addons.go:101
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).Run.func1
	cmd/kubeadm/app/cmd/phases/workflow/runner.go:234
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).visitAll
```

try:
```
securityContext:
      capabilities:
        add: ["NET_ADMIN"]
```

Comment (kevensen) [reply cmd/kubeadm/app/phases/addons/proxy/manifests.go:85]: Drat.  Ok.  I'll be able to try that in a couple hours

**Comment by kevensen** [comment]
I signed it

File: cluster/addons/kube-proxy/kube-proxy-ds.yaml [hunk 3]
```diff
@@ -48,7 +48,8 @@ spec:
         {{kube_cache_mutation_detector_env_name}}
```
Comment (liggitt) [inline cluster/addons/kube-proxy/kube-proxy-ds.yaml:48]: note the comment at the top about keeping this in sync with kube-proxy.manifest

Comment (kevensen) [reply cluster/addons/kube-proxy/kube-proxy-ds.yaml:48]: Thanks for the tip.  Unfortunately I am not seeing the file to which that is referring.

Comment (cmluciano) [reply cluster/addons/kube-proxy/kube-proxy-ds.yaml:48]: Should be here https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/manifests/kube-proxy.manifest

**Review by cmluciano — COMMENTED** [review]
I'm trying to figure out if there is an open issue related to this work. I believe there was always hope for the `privileged` flag to be removed but NET_ADMIN might not be enough. KubeProxy needs capabilities to load found kernel modules on demand and that kind of sounds like a SYS_ADMIN capability but I'm not 100% on the division of boundaries.

**Comment by cmluciano** [comment]
cc @kubernetes/sig-network-misc

**Comment by neolit123** [comment]
@kevensen the kube-proxy YAML for kubeadm still seems invalid.
you could build kubeadm locally and test `kubeadm init` before committing the fixed YAML.

**Comment by bowei** [comment]
/assign

Reading through the privileges of CAP_SYS_ADMIN seems to allow a ton of stuff. Is it not possible to subset?

**Comment by kevensen** [comment]
I added CAP_SYS_ADMIN because I was trying to make this pass the tests and was thinking that permissions were an issue.  I plan to remove it once I figure out the manifest format issue.

**Comment by neolit123** [comment]
you can see kube-proxy pod failing on worker nodes here:

summary:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/96280/pull-kubernetes-e2e-kind/1325191254480261120

artifacts:
https://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/96280/pull-kubernetes-e2e-kind/1325191254480261120/artifacts/logs/kind-worker/containers/

kube-proxy log:
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/96280/pull-kubernetes-e2e-kind/1325191254480261120/artifacts/logs/kind-worker/containers/kube-proxy-6l7dv_kube-system_kube-proxy-7004be78950041cd29e5e07649839298efe44b0201af10251c982915be603e54.log

> 2020-11-07T22:17:44.508452778Z stderr F F1107 22:17:44.508405       1 server.go:495] open /proc/sys/net/netfilter/nf_conntrack_tcp_timeout_established: read-only file system

**Comment by kevensen** [comment]
Ok.  After doing a bit of research and recollecting how `/proc` works, it is generally inaccessible by a container not running as privileged.  A few of the Linux capabiltiies allow for tweaking the values under `/proc`.  But two values being referenced in the logs are not covered by any of the capabilities.

* /proc/self/oom_score_adj
* /proc/sys/net/netfilter/nf_conntrack_tcp_timeout_established

I am resigned to conclude that the kube-proxy must be run as privileged.  I appreciate everyone's help and suggestions.

**Comment by neolit123** [comment]
Thank you for trying! Hope one day we can limit the kube-proxy priviliges,
somehow.

I am going to go ahead and close this, but if others have comments please
go ahead.

/close

**Pull request closed without merging**